torch/nn/parallel/replicate.py

```diff
@@ -83,26 +106,124 @@ def _copy_scriptmodule_methods(modules, module_copies, module_indices):
             replica._copy_method(method_name, param_list, module)
 
 
+# group tensors on the same device together, which can later be broadcast to
+# a list of devices. For example,consider 5 tensors on 2 devices
+#   a = torch.Tensor(0).cuda(0)
+#   b = torch.Tensor(0).cuda(0)
+#   c = torch.Tensor(0).cuda(1)
+#   d = torch.Tensor(0).cuda(0)
+#   e = torch.Tensor(0).cuda(1).
+# Let inputs be
+#   tensors = [a, b, c, d, e] and
+#   devices = [[0, 1], [2, 3]].
+# Then, outputs will be:
+#   grouped_tensors = [[a, b, d], [c, e]],
+#   grouped_devices = [[0, 2], [1, 3]],
+#   original_index = [[0, 1, 3], [2, 4]],
+# meaning that grouped_tensors[i] will be broadcast to grouped_devices[i].
+def _group_by_device(tensors, devices):
+    if isinstance(devices[0], list):
+        # device id to output group index, this is necessary when `tensors` only
+        # use a subset of devices in `devices[0]`
+        dev_to_idx = {dev: i for i, dev in enumerate(
+            set([t.device.index for t in tensors]))}
```

Comment:
It took me a while to figure out how this would work, since you're iterating over a set (unordered). I think it is possible for the groups to be flipped from the input ordering because of that (is that true, or do Python sets still have some ordering?).

Possible alternative (which should reflect ordering of devices of input tensors):

```python
device_to_group_index = {}
for t in tensors:
  if not t.device.index in device_to_group_index:
    device_to_group_index[t.device.index] = len(device_to_group_index)
```

Reply:
Is maintaining the order necessary here? `original_index` on line 138 remembers `tensors`' original order, and `_broadcast_coalesced_reshape ` will recover the orders

Reply:
suggested changes applied, for readability and easier tests :)